Throw exceptions when unrecoveral error situation occur

MKodde opened the pull request:
The library was very tolerant when error situations where encountered. Some of these situations would result in an unrecoverable situation. For example. When during generation of the enrollment key, your enrollment key can not be stored, the app would churn on. 

The consumer (Tiqr-gssp) would not be able to discern when an error occurs. 

By throwing exceptions in these situations, the predictability increases greatly. Allowing the consumer of the library to decide on how to deal with those situations.

https://www.pivotaltracker.com/story/show/181553909

MKodde committed:
Create the ReadWriteException

MKodde committed:
Throw ReadWrite exception in StateStorage

MKodde committed:
Throw ReadWrite Exception in UserSecretStorage

MKodde committed:
Throw ReadWrite Exception in UserStorage

MKodde committed:
Rename ReadWriteException

Removing the ZendLike namespace from the classname

VadimSchmitz commented on library/tiqr/Tiqr/UserSecretStorage.php:
```diff
@@ -73,8 +73,16 @@ public static function getSecretStorage(string $type = "file", LoggerInterface $
+
                 require_once("Tiqr/UserSecretStorage/Pdo.php");
-                return new Tiqr_UserSecretStorage_Pdo($encryption, $logger, $dsn, $userName, $password, $tableName);
+                return new Tiqr_UserSecretStorage_Pdo($encryption, $logger, $handle, $tableName);
```
very nice

VadimSchmitz reviewed (commented):
Unfortunately i cant seem to spot the bugs that you have encountered

VadimSchmitz commented on library/tiqr/Tiqr/StateStorage/Pdo.php:
```diff
@@ -68,12 +68,13 @@ private function keyExists($key)
-        if (!$sth->execute(array(time()))){
+        $result = $sth->execute(array(time()));
+        if (!$result || $sth->rowCount() === 0){
             $this->logger->error('Unable to remove expired keys from the pdo state storage');
```
Is there a reason for the normal log instead of an exception?

MKodde replied:
As we dicussed IRL (but stated here for the record)

Given this method is not disturbing the process (it is called to clean up expired records). It is not a problem when it fails. The next call should perform the same task again. No data is lost, the application can continue without issue.

I will add a comment to explain this in code

MKodde committed:
Adjust the PDO unit test

The exception is no longer thrown, the test neeed to be adjusted
accordingly.

While at it. I updated the formatting of the test methods slightly.
